token-lending/program/src/processor.rs

```diff
@@ -1571,94 +1577,92 @@ fn process_flash_loan(program_id: &Pubkey, amount: u64, accounts: &[AccountInfo]
-        source: source_liquidity_info.clone(),
-        destination: destination_liquidity_info.clone(),
-        amount: flash_loan_amount,
-        authority: derived_lending_market_account_info.clone(),
-        authority_signer_seeds,
-        token_program: token_program_id.clone(),
-    })?;
-
-    let mut data = Vec::with_capacity(RECEIVE_FLASH_LOAN_INSTRUCTION_DATA_SIZE);
-    data.push(RECEIVE_FLASH_LOAN_INSTRUCTION_TAG);
-    data.extend_from_slice(&returned_amount_required.to_le_bytes());
-
     let mut flash_loan_instruction_accounts = vec![
         AccountMeta::new(*destination_liquidity_info.key, false),
         AccountMeta::new(*source_liquidity_info.key, false),
         AccountMeta::new_readonly(*token_program_id.key, false),
     ];
-
     let mut flash_loan_instruction_account_infos = vec![
         destination_liquidity_info.clone(),
-        flash_loan_receiver_program_info.clone(),
+        flash_loan_receiver_program_id.clone(),
         source_liquidity_info.clone(),
         token_program_id.clone(),
     ];
-
-    for flash_loan_receiver_additional_account_info in account_info_iter {
-        flash_loan_instruction_account_infos
-            .push(flash_loan_receiver_additional_account_info.clone());
+    for account_info in account_info_iter {
         flash_loan_instruction_accounts.push(AccountMeta {
-            pubkey: *flash_loan_receiver_additional_account_info.key,
-            is_signer: flash_loan_receiver_additional_account_info.is_signer,
-            is_writable: flash_loan_receiver_additional_account_info.is_writable,
+            pubkey: *account_info.key,
+            is_signer: account_info.is_signer,
+            is_writable: account_info.is_writable,
         });
+        flash_loan_instruction_account_infos.push(account_info.clone());
     }
 
-    let flash_loan_receiver_instruction = Instruction {
-        program_id: *flash_loan_receiver_program_info.key,
-        accounts: flash_loan_instruction_accounts,
-        data,
-    };
+    reserve.liquidity.borrow(flash_loan_amount_decimal)?;
+    Reserve::pack(reserve, &mut reserve_info.data.borrow_mut())?;
+
+    spl_token_transfer(TokenTransferParams {
+        source: source_liquidity_info.clone(),
+        destination: destination_liquidity_info.clone(),
+        amount: flash_loan_amount,
+        authority: lending_market_authority_info.clone(),
+        authority_signer_seeds,
+        token_program: token_program_id.clone(),
+    })?;
+
+    const RECEIVE_FLASH_LOAN_INSTRUCTION_DATA_SIZE: usize = 9;
+    const RECEIVE_FLASH_LOAN_INSTRUCTION_TAG: u8 = 0u8;
+
+    let mut data = Vec::with_capacity(RECEIVE_FLASH_LOAN_INSTRUCTION_DATA_SIZE);
+    data.push(RECEIVE_FLASH_LOAN_INSTRUCTION_TAG);
+    data.extend_from_slice(&returned_amount_required.to_le_bytes());
+
     invoke(
-        &flash_loan_receiver_instruction,
+        &Instruction {
+            program_id: *flash_loan_receiver_program_id.key,
```

Comment:
To be safe, let's make sure that this program id is not the same as the lending program id to avoid reentrancy attacks

Reply:
Done.